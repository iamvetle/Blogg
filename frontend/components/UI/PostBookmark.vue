<template>
    <div>
        <BaseIconSaveArticleSaved v-if="checkIfPostIsSaved(post)" @click="unsave(post)" />

        <BaseIconSaveArticleUnSaved v-else @mouseover="color = 'fill-primary'" @mouseleave="color = 'fill-black'"
            @click="save(post)" :fill-color="color" />
    </div>
</template>

<script setup lang="ts">
/**
 * This component handles the functionality for saving and unsaving articles. 
 * It shows different icons based on the saved state of the article.
 */

// Props definition
// @prop post - Numeric ID of the article.
defineProps<{
    post: number;
}>();

// Reactive variable for icon color
const color = ref("fill-black");

// Importing the loggedInUserStore from the store directory
import { useLoggedInUserStore } from '~/store/loggedInUserStore';

// Using the loggedInUserStore for managing the saved posts
const loggedInUserStore = useLoggedInUserStore();

/**
 * Removes the selected post from the saved posts list.
 * @param post - Numeric ID of the post to unsave.
 */
const unsave = async (post: number) => {
    const index = loggedInUserStore.idArrayOfSavedPosts.findIndex((id) => id === post);
    loggedInUserStore.idArrayOfSavedPosts.splice(index, 1);
    await getSaveOrUnsavePost(post);
}

/**
 * Adds the selected post to the saved posts list.
 * @param post - Numeric ID of the post to save.
 */
const save = async (post: number) => {
    await getSaveOrUnsavePost(post);
}

// Note: The function getSaveOrUnsavePost is used but not defined or imported in this snippet.
// Ensure it is properly defined and imported for the component to function correctly.
</script>

<style scoped></style>