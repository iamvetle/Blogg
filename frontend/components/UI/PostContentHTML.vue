<template>
    <div data-test="html-content" v-bind="$attrs" v-html="html"></div>
</template>

<script setup lang="ts">
/**
 * Renders the prop element - which is supposed to be 
 * html content as one big string
 */
defineProps<{
    html: string;
}>();

</script>

<style scoped></style>