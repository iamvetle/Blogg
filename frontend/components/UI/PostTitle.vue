<template>
	<h1 v-bind="$attrs">{{ title }}</h1>
</template>

<script setup lang="ts">
/**
 * Displays the text prop
 */

defineProps<{
	title: string;
}>();
</script>

<style scoped></style>
