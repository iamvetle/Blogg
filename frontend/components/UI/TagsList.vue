<template>
    <div v-if="tags" class="mb-2 flex space-x-1" id="tags">
        <span class="flex" v-for="tag in tags">
            <BaseTag :text="tag" :key="tag.name" />
        </span>
    </div>
</template>

<script setup lang="ts">
/**
 * Contains multiple tags.
 * 
 * Loops through the tags, one after another.
 */

defineProps<({
    tags: TagType[]
})>()

</script>

<style scoped></style>