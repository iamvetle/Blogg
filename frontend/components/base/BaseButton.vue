<template>
    <button data-test="button">{{ text }}</button>
</template>

<script setup lang="ts">

defineProps<{
    text: string
}>()

</script>

<style scoped></style>