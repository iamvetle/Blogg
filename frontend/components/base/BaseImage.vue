<template>
    <div>
        <img :src="src" v-bind="$attrs"/>
    </div>
</template>

<script setup lang="ts">

defineProps<{
    src:string;
}>();

</script>

<style scoped>

</style>