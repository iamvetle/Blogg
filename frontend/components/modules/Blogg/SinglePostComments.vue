<template>
    <div v-if="postStore.allComments">
        <h2>Comments written: ({{ num_of_comments }})</h2>
        <div>
            <SinglePostCommentsList :comments="postStore.allComments" :post-author="post.author.username"/>
        </div>

        <hr>

        <SinglePostCommentAdd v-if="authStore.isAuthenticated" :post-id="post.id" />
    </div>
</template>

<script setup lang="ts">

/**
 * This component is part of the single post [id] page
 * It deals with the comments associated with the post and
 * other comments stuff
 */

const authStore = useAuthStore()

const postStore = usePostStore()

/**
 * Counts the number of comments - I do it this way for me to keep the reactivity
 */
const num_of_comments = computed(() => {
    return postStore.allComments?.length ?? 0

})

defineProps<{
    post:PostSingleType
}>();

</script>

<style scoped></style>