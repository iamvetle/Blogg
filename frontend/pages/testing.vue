<template>
  <div class="flex">

    <form @submit.prevent method="post" class="flex flex-col" enctype="multipart/form-data">

      <label for="name">Name:</label>
      <input type="text" name="name" id="name">

      <label for="email">Email</label>
      <input type="email" name="email" id="email">

      <label for="message">Message:</label>
      <textarea name="message" id="" cols="30" rows="10" placeholder="Write a message here"></textarea>

      <label for="file">File:</label>
      <input type="file" name="file" id="file">

      <button type="submit">Submit</button>
    </form>

  </div>
</template>

<script setup lang="ts">
onMounted(() => {

  const form = document.querySelector("form")

  form?.addEventListener("submit", (e: any) => {


    const formData = new FormData(form)

    for (let item of formData) {
      console.log(item[0], item[1])
    }
    
    // ta axios of prÃ¸v fetch her
  })





})

// watchEffect(() => {

// })

</script>

<style scoped></style>