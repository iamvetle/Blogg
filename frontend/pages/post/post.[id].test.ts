import idVue from './[id].vue'
import { flushPromises, shallowMount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

// Components in the page
import SinglePostComments from '~/components/modules/Blogg/SinglePostComments.vue';
import PostBookmark from '~/components/UI/PostBookmark.vue';
import TagsList from '~/components/UI/TagsList.vue';
import PostTitle from '~/components/UI/PostTitle.vue';
import BaseImage from '~/components/base/BaseImage.vue';
import PostContentHTML from '~/components/UI/PostContentHTML.vue';
import BaseButton from '~/components/base/BaseButton.vue';


const standardPost = {
    id: 5,
    title: "testTitle",
    content: "testContent",
    author: {
        username: "testUsername",
        first_name: "testFirstName",
        last_name: "testLastName"
    },
    num_of_comments: 6,
    tags: [
        {
            name: "candy",
        },
        {
            name: "fridge"
        }
    ]
}

let wrapper: any;
let pinia: any = createTestingPinia();

// let generalStore; 
let postStore;
// let loggedInUserStore; 
// let paginationStore; 

const mockFetchPostRelated = vi.fn()

const factory = () => {
    return shallowMount(idVue, {
        global: {
            plugins: [pinia],
            components: {
                SinglePostComments,
                PostBookmark,
                TagsList,
                PostTitle,
                BaseImage,
                PostContentHTML,
                BaseButton,
            },
            mocks: {
                fetchPostRelated:mockFetchPostRelated
            },
            stubs: {
                TagsList: true,
                SinglePostComments:true,
                BaseButton:true,
                PostContentHTML:true,
                PostTitle:true,
                PostBookmark:true,
                BaseImage:true
            },
        },
        props: {},
        slots: {}
    })
};

describe('Testing the single post component', () => {

    vi.stubGlobal('definePageMeta', () => {
        return null
    });

    // vi.stubGlobal('fetchPostRelated', () => {
    //     return null
    // });

    vi.stubGlobal('useRoute', () => {
        return {
            params: {
                id: 1
            }
        }
    });

    beforeEach(() => {
        // generalStore = useGeneralStore(pinia); 
        postStore = usePostStore(pinia);
        // loggedInUserStore = useLoggedInUserStore(pinia); 
        // paginationStore = usePaginationStore(pinia); 

    });

    afterEach(() => {
        if (wrapper) {
            wrapper.unmount();
        }
    });

    test('Page should exist', () => {
        wrapper = factory()

        expect(wrapper.exists()).toBe(true)
    })
    test('Should match snapshot', () => {
        wrapper = factory()

        expect(wrapper).toMatchSnapshot()
    })

    test('PostBookmark should be present if all post data is ok', async () => {
        wrapper = factory()

        wrapper.vm.post = standardPost

        await wrapper.vm.$nextTick()

        const bookmark = wrapper.findComponent({ name: "PostBookmark" })
        expect(bookmark.exists()).toBe(true)
    })
    test('articletags should be present if all post data is ok', async () => {
        wrapper = factory()

        wrapper.vm.post = standardPost

        await wrapper.vm.$nextTick()

        const bookmark = wrapper.findComponent({ name: "TagsList" })
        expect(bookmark.exists()).toBe(true)
    })
    test('Should have the image component', async () => {
        wrapper = factory()
        wrapper.vm.post = standardPost
        await wrapper.vm.$nextTick()
        expect(wrapper.findComponent({ name: "BaseImage" }).exists()).toBe(true)
    })
    test('Should have the component that renders the html input from prop', async () => {
        wrapper = factory()
        wrapper.vm.post = standardPost
        await wrapper.vm.$nextTick()
        console.log(wrapper.html())
        expect(wrapper.findComponent({ name: "PostContentHTML" }).exists()).toBe(true)
    })  
    test('Should render the basebutton, back button, on page', async () => {
        wrapper = factory()
        wrapper.vm.post = standardPost
        await wrapper.vm.$nextTick()
        expect(wrapper.findComponent({ name: "BaseButton" }).exists()).toBe(true)
    })
    test('Should render the comments component', async () => {
        wrapper = factory()
        
        wrapper.vm.post = standardPost

        await wrapper.vm.$nextTick()
        console.log(wrapper.html())
    
        expect(wrapper.find("[data-test='comments']").exists()).toBe(true)
        expect(wrapper.findComponent({ name:"SinglePostComments" }).exists()).toBe(true)


    })
    test('Should render the post title component', async () => {
        wrapper = factory()
        
        wrapper.vm.post = standardPost

        await wrapper.vm.$nextTick()
        console.log(wrapper.html())
    
        expect(wrapper.find("[data-test='post_title']").exists()).toBe(true)
        expect(wrapper.findComponent({ name:"PostTitle" }).exists()).toBe(true)
    })


})