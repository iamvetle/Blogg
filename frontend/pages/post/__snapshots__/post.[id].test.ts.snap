// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the single post component > Should match snapshot 1`] = `"<!--v-if-->"`;

exports[`Testing the single post component > Should match snapshot 2`] = `"<!--v-if-->"`;
